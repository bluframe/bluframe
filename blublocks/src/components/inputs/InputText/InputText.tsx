/* eslint-disable no-ternary, max-lines-per-function */

import {
  ErrorText,
  InfoText,
  Input,
  Label,
  LoadingWrapper,
  SuccessText,
  Wrapper
} from "./styled"
import { ForwardedRef, forwardRef } from "react"
import { ComponentProps } from "."
import Loading from "components/Loading"

const InputText = forwardRef(
  (
    {
      className,
      defaultValue,
      disabled,
      error,
      errorText,
      hasContent,
      infoText,
      inputId,
      inputProps,
      isFullBorder,
      label,
      labelId,
      loading,
      max,
      min,
      name,
      onBlur,
      onChange,
      onInput,
      onKeyDown,
      required,
      size,
      step,
      success,
      successText,
      type = "text",
      value
    }: ComponentProps,
    ref: ForwardedRef<HTMLInputElement>
  ) => (
    <Wrapper className={className}>
      <Input
        {...inputProps}
        $error={error}
        $hasContent={hasContent}
        $isFullBorder={isFullBorder}
        $loading={loading}
        $success={success}
        aria-invalid={error ? "true" : "false"}
        defaultValue={defaultValue}
        disabled={disabled}
        id={inputId}
        max={max}
        min={min}
        name={name}
        onBlur={onBlur}
        onChange={onChange}
        onInput={onInput}
        onKeyDown={onKeyDown}
        ref={ref}
        required={required}
        size={size}
        step={step}
        type={type}
        value={value}
      />
      <Label htmlFor={inputId} id={labelId} isFullBorder={isFullBorder}>
        {label}
      </Label>

      {loading ? (
        <LoadingWrapper>
          <Loading />
        </LoadingWrapper>
      ) : null}

      {errorText ? <ErrorText role="alert">{errorText}</ErrorText> : null}
      {infoText ? <InfoText aria-live="polite">{infoText}</InfoText> : null}
      {successText ? (
        <SuccessText aria-live="polite">{successText}</SuccessText>
      ) : null}
    </Wrapper>
  )
)

InputText.displayName = "InputText"

export default InputText
