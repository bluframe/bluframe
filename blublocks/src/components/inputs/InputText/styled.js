// @flow
/* eslint-disable no-ternary */

import styled, { type StyledComponent } from "styled-components"
import type { Theme } from "@bluframe/blublocks"

export const Wrapper: StyledComponent<{||}, Theme, HTMLDivElement> = styled.div`
  display: flex;
  flex-direction: column;
`

export const Input: StyledComponent<
  {||},
  Theme,
  HTMLInputElement
> = styled.input.attrs({ type: "text" })`
  border: 1px solid ${({ theme }) => theme.palette.primary.main};
  border-radius: 4px;
  font-size: 16px;
  padding: 8px;
  width: 100%;
`

export const Label: StyledComponent<
  {||},
  Theme,
  HTMLLabelElement
> = styled.label`
  color: ${({ theme }) => theme.palette.text.main};
  font-size: 16px;
  margin-bottom: 4px;
`
