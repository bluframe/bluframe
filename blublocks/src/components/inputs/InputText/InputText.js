// @flow
/* eslint-disable no-ternary, react/jsx-no-bind */

import { Input, Label, Wrapper } from "./styled"
import type { ComponentProps } from "."
import React from "react"

const InputText = ({
  inputId,
  label,
  labelId,
  name,
  onChange,
  value
}: ComponentProps): React$Node => (
  <Wrapper>
    <Label htmlFor={inputId} id={labelId}>
      {label}
    </Label>
    <Input
      id={inputId}
      name={name}
      onChange={(event) => onChange(event.target.value)}
      value={value}
    />
  </Wrapper>
)

export default InputText
