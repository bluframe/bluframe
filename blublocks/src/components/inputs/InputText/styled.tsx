/* eslint-disable no-ternary, no-nested-ternary */

import styled, { css } from "styled-components"
import { Theme } from "styles/theme"

export const Wrapper = styled.div`
  display: grid;
  position: relative;
  row-gap: 4px;
  width: 100%;
`

const labelShrinkCss = css`
  font-size: 12px;
  left: 16px;
  top: ${({ isFullBorder }: { isFullBorder?: boolean }) =>
    isFullBorder ? "-24px" : "0"};
  transition: all 200ms;

  & svg {
    height: 12px;
    width: 12px;
  }
`

export const Input = styled.input.attrs({ placeholder: " " })<{
  $error?: boolean
  $hasContent: boolean
  $success?: boolean
  isFullBorder?: boolean
}>`
  ${({ $error, $success, isFullBorder, theme }) =>
    isFullBorder
      ? `border: 1px solid ${
          $error
            ? theme.palette.error.main
            : $success
            ? theme.palette.success.main
            : theme.colors.gray.light
        }`
      : `
  border: none;
border-bottom: 1px solid ${
          $error
            ? theme.palette.error.main
            : $success
            ? theme.palette.success.main
            : theme.colors.gray.light
        };
  `};
  border-radius: 4px;
  box-shadow: none;
  color: ${({ disabled, theme }: { disabled?: boolean; theme: Theme }) =>
    disabled ? theme.palette.text.light : theme.palette.text.main};
  font-size: 16px;
  height: ${({ isFullBorder }) => (isFullBorder ? "44px" : "50px")};
  margin: 0;
  padding: ${({ isFullBorder }) => (isFullBorder ? "8px 16px" : "16px 16px 0")};
  width: 100%;

  &:focus {
    ${({ isFullBorder, theme }) =>
      isFullBorder
        ? `border: 1px solid ${theme.palette.primary.main}`
        : `
  border: none;
border-bottom: 1px solid ${theme.palette.primary.main};
  `};
    box-shadow: none;
  }

  &:not(:placeholder-shown) + label,
  &:focus + label {
    ${labelShrinkCss}
  }

  ${({ $hasContent }: { $hasContent: boolean }) =>
    $hasContent &&
    css`
      & + label {
        ${labelShrinkCss}
      }
    `}
`

export const Label = styled.label<{ isFullBorder?: boolean }>`
  color: ${({ theme }: { theme: Theme }) => theme.palette.text.light};
  font-size: 16px;
  left: 16px;
  pointer-events: none;
  position: absolute;
  top: ${({ isFullBorder }: { isFullBorder?: boolean }) =>
    isFullBorder ? "8px" : "12px"};
  transition: all 200ms;
`

export const ErrorText = styled.div`
  color: ${({ theme }: { theme: Theme }) => theme.palette.error.main};
  font-size: 12px;
`

export const InfoText = styled.div`
  color: ${({ theme }: { theme: Theme }) => theme.palette.info.main};
  font-size: 12px;
`

export const SuccessText = styled.div`
  color: ${({ theme }: { theme: Theme }) => theme.palette.success.main};
  font-size: 12px;
`
