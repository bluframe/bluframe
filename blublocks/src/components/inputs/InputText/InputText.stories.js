// @flow

import InputText, { type Props } from "components/inputs/InputText"
import React from "react"
import type { Story } from "@storybook/react"
import { action } from "@storybook/addon-actions"

const stories = {
  component: InputText,
  title: "Inputs/Text Input"
}

const Template = (args: Props) => <InputText {...args} />

const args = {
  inputId: "name-input",
  label: "Name",
  name: "name",
  onChange: action("onChange"),
  value: ""
}

export const Default: Story<Props> = Template.bind({})
Default.args = args

export default stories
