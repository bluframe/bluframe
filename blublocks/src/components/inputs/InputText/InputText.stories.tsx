import Icon from "components/Icon"
import InputText from "components/inputs/InputText"
import React from "react"
import { StoryObj } from "@storybook/react"
import { action } from "@storybook/addon-actions"

const meta = {
  component: InputText,
  title: "Inputs/Text Input"
}

type Story = StoryObj<typeof InputText>

const args = {
  inputId: "name-input",
  label: "Name",
  name: "name",
  onBlur: action("onBlur"),
  onChange: action("onChange")
}

export const Default: Story = {
  args
}

export const DefaultValue: Story = {
  args: {
    ...args,
    defaultValue: "Edoardo"
  }
}

export const Value: Story = {
  args: {
    ...args,
    value: "Edoardo"
  }
}

export const FullBorder: Story = {
  args: {
    ...args,
    isFullBorder: true
  }
}

export const Info: Story = {
  args: {
    ...args,
    infoText: "You can only set your name once!"
  }
}

export const InfoFullBorder: Story = {
  args: {
    ...args,
    infoText: "You can only set your name once!",
    isFullBorder: true
  }
}

export const Success: Story = {
  args: {
    ...args,
    success: true,
    successText: "Username is available!"
  }
}

export const SuccessFullBorder: Story = {
  args: {
    ...args,
    isFullBorder: true,
    success: true,
    successText: "Username is available!"
  }
}

export const ErrorAndMessage: Story = {
  args: {
    ...args,
    error: true,
    errorText: "This field is required"
  }
}

export const ErrorFullBorder: Story = {
  args: {
    ...args,
    error: true,
    errorText: "This field is required",
    isFullBorder: true
  }
}

export const Required: Story = {
  args: {
    ...args,
    required: true
  }
}

export const NumberType: Story = {
  args: {
    ...args,
    type: "number"
  }
}

export const NumberMaxMin: Story = {
  args: {
    ...args,
    max: 100,
    min: 0,
    type: "number"
  }
}

export const Disabled: Story = {
  args: {
    ...args,
    disabled: true
  }
}

export const LabelIcon: Story = {
  args: {
    ...args,
    label: (
      <>
        <Icon group="io5" name="IoPerson" />
        {"Name"}
      </>
    )
  }
}

export default meta
