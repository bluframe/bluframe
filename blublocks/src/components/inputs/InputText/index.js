// @flow

import InputText from "./InputText"
import type { InputTextProps } from "@bluframe/blublocks"
import { prepareComponent } from "@bluframe/grapple"

export type Props = {|
  ...InputTextProps
|}

export type ComponentProps = {|
  ...Props,
  +labelId?: string
|}

const usePrepareComponent = (props: Props): ComponentProps => {
  const labelId = props.inputId && `${props.inputId}-label`

  return {
    ...props,
    labelId
  }
}

export default (prepareComponent<Props, ComponentProps>(usePrepareComponent)(
  InputText
): React$ComponentType<Props>)
