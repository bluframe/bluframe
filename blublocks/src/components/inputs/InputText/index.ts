"use client"

import React, {
  ChangeEvent,
  FocusEvent,
  FormEventHandler,
  ForwardedRef,
  KeyboardEvent,
  useState
} from "react"
import InputText from "./InputText"
import { prepareComponent } from "@bluframe/grapple"

export interface Props {
  className?: string
  defaultValue?: string
  disabled?: boolean
  error?: boolean
  errorText?: string
  infoText?: string
  inputId?: string
  inputProps?: {
    "aria-activedescendant"?: string
    "aria-autocomplete"?: "list" | "none" | "inline" | "both"
    "aria-controls"?: string
    "aria-expanded"?: boolean
    "aria-labelledby"?: string
    autoComplete?: string
    id?: string
    // eslint-disable-next-line no-unused-vars
    onBlur?: (event: FocusEvent) => void
    // eslint-disable-next-line no-unused-vars
    onChange?: (event: ChangeEvent) => void
    role?: string
  }
  isFullBorder?: boolean
  label: React.ReactNode
  loading?: boolean
  max?: number
  min?: number
  // eslint-disable-next-line no-unused-vars
  onBlur?: (event: FocusEvent) => void
  // eslint-disable-next-line no-unused-vars
  onChange?: (event: ChangeEvent) => void
  // eslint-disable-next-line no-unused-vars
  onKeyDown?: (event: KeyboardEvent) => void
  name: string
  ref?: ForwardedRef<HTMLInputElement>
  required?: boolean
  size?: number
  step?: string
  success?: boolean
  successText?: string
  type?: string
  value?: string
}

export interface ComponentProps extends Props {
  hasContent: boolean
  labelId?: string
  onInput: FormEventHandler<HTMLInputElement>
}

const ZERO = 0

const usePrepareComponent = (props: Props): ComponentProps => {
  const labelId = props.inputId && `${props.inputId}-label`
  const [hasValue, setHasValue] = useState<boolean | null>(null)

  const onBlur = (event: FocusEvent) => {
    if (props.onBlur) {
      props.onBlur(event)
    }

    if (props.inputProps?.onBlur) {
      props.inputProps.onBlur(event)
    }
  }

  const onChange = (event: ChangeEvent) => {
    if (props.onChange) {
      props.onChange(event)
    }

    if (props.inputProps?.onChange) {
      props.inputProps.onChange(event)
    }
  }

  const onInput: FormEventHandler<HTMLInputElement> = (event) => {
    // @ts-ignore
    setHasValue(event.target.value.length > ZERO)
  }

  const hasContent = Boolean(hasValue ?? props.defaultValue ?? props.value)

  return {
    ...props,
    hasContent,
    labelId,
    onBlur,
    onChange,
    onInput
  }
}

export default prepareComponent<Props, ComponentProps>(usePrepareComponent)(
  InputText
)
