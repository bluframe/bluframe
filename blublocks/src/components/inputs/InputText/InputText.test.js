// @flow

import type { ComponentProps } from "."
import InputText from "./InputText"
import React from "react"
import { render } from "tests"

jest.mock(".", () => ({}))

describe("InputText", () => {
  const props: ComponentProps = {
    inputId: "name-input",
    label: "Name",
    name: "name",
    onChange: jest.fn(),
    value: ""
  }

  it("renders", () => {
    const { container } = render(<InputText {...props} />)

    expect(container.firstChild).toMatchSnapshot()
  })
})
