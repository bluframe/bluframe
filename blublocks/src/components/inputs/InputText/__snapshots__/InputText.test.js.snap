// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputText renders 1`] = `
<div
  class="styled__Wrapper-sc-1svvkyj-0 eHUKlz"
>
  <label
    class="styled__Label-sc-1svvkyj-2 kfTdUg"
    for="name-input"
  >
    Name
  </label>
  <input
    class="styled__Input-sc-1svvkyj-1 hBPipW"
    id="name-input"
    name="name"
    type="text"
    value=""
  />
</div>
`;
